eip_operator/src/controller/pod.rs

```diff
@@ -33,60 +33,88 @@ impl k8s_controller::Context for Context {
-            Some(eni_id) => eni_id,
-            None => {
-                let instance_description =
-                    crate::aws::describe_instance(&self.ec2_client, instance_id).await?;
-
-                crate::aws::get_eni_from_private_ip(&instance_description, pod_ip)
-                    .ok_or(Error::NoInterfaceWithThatIp)?
+        let eips = eip_api.list(&ListParams::default()).await?.items;
+        // Cases
+        // there are some eips
+        // - we are associated with one
+        //   - it is correct
+        //     - done - register dns labels - ok
+        //   - it is incorrect
+        //     - ask eip to reoncile, reschedule
+        // - we are not associated with any
+        //   - we want to be
+        //     - an eip is not associated (potentially free to grab)
+        //       - we ask to claim it
+        //     - it is associated (someone else is using it)
+        //       - done ok
+        //   - we do not want to be
+        //     - done ok
+        // no eips
+        // - done ok
+        if !eips.is_empty() {
+            // there are eips
+            if let Some(eip) = eips.iter().find(|eip| eip.associated_with_pod(pod)) {
+                // we are associated with an eip
+                if eip.matches_pod_name(&pod.name_unchecked()) {
+                    // we are associated, it is correct
+                    info!(
+                        "Pod {} is already attached to EIP {}, setting up DNS target annotation",
+                        pod.name_unchecked(),
+                        eip.name_unchecked()
+                    );
+                    let allocation_id = eip.allocation_id().ok_or(Error::MissingAllocationId)?;
+                    let eip_description =
+                        crate::aws::describe_address(&self.ec2_client, allocation_id)
+                            .await?
+                            .addresses
+                            .ok_or(Error::MissingAddresses)?
+                            .swap_remove(0);
+                    let public_ip = eip_description.public_ip.ok_or(Error::MissingPublicIp)?;
+                    add_dns_target_annotation(&pod_api, name, &public_ip, allocation_id).await?;
+                    return Ok(None);
+                } else {
+                    // we are associated, it is incorrect
+                    info!("Pod {} is attached to EIP {}, but should not be, asking eip to reconcile, and rescheduling reconciliation", pod.name_unchecked(), eip.name_unchecked());
+                    crate::eip::trigger_reconciliation(&eip_api, &eip.name_unchecked()).await?;
+                    return Ok(Some(Action::requeue(Duration::from_secs(2))));
+                }
+            } else {
+                // we are not associated
+                if let Some(eip) = eips
+                    .iter()
+                    .find(|eip| eip.matches_pod_name(&pod.name_unchecked()) && !eip.attached())
```

Comment:
I wonder about the `!eip.attached()` part. EIPs with pod matchers may only match a single pod, so if it is attached, but not associated with this pod, that means it is incorrectly associated and should be re-reconciled.